python/cudf_polars/cudf_polars/experimental/fusion.py

```diff
@@ -0,0 +1,187 @@
+                ir.schema, (*child.subnodes, ir), child.fused_io, *grandchildren
+            )
+        else:
+            new_node = Fused(ir.schema, (ir,), None, *children)
+    else:
+        new_node = ir.reconstruct(children)
+
+    partition_info[new_node] = rec.state["partition_info"][ir]
+    return new_node, partition_info
+
+
+def _is_fusible_ir_node(
+    ir: IR, partition_info: MutableMapping[IR, PartitionInfo]
+) -> bool:
+    return (
+        # Basic fusion.
+        # These are partition-wise operations if they exist
+        # in the graph after lowering (the data-movement
+        # has been encoded with Shuffle/Reduction/etc).
+        isinstance(
+            ir,
+            (Cache, Filter, GroupBy, HStack, MapFunction, Projection, Select),
+        )
+        or (
+            # IO Fusion.
+            # Multi-partition IO is always implemented as a Union
+            # over Scan/SplitScan nodes. The number of children
+            # must correspond to the number of partitions.
+            isinstance(ir, Union)
+            and all(isinstance(n, (Scan, SplitScan)) for n in ir.children)
+            and partition_info[ir].count == len(ir.children)
+        )
+    )
+
+
+@generate_ir_tasks.register(Fused)
+def _(
+    ir: Fused, partition_info: MutableMapping[IR, PartitionInfo]
+) -> MutableMapping[Any, Any]:
+    from cudf_polars.experimental.parallel import _default_generate_ir_tasks
+
+    if ir.fused_io is None:
+        return _default_generate_ir_tasks(ir, partition_info)
+
+    graph: MutableMapping[Any, Any] = {}
+    for i, key in enumerate(partition_info[ir].keys(ir)):
+        io = ir.fused_io.children[i]
+        assert isinstance(io, (Scan, SplitScan))
+        graph[key] = (
+            ir.do_evaluate,
+            ir._non_child_args[0],
+            ir._non_child_args[1],
+            [io.do_evaluate],
+            list(io._non_child_args),
+        )
+    return graph
+
+
+def fuse_ir_graph(
+    ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo],
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    """
+    Rewrite an IR graph with fused nodes.
+
+    Parameters
+    ----------
+    ir
+        Root of the lowered-IR graph to rewrite.
+    partition_info
+        Initial partitioning information.
+
+    Returns
+    -------
+    new_ir, partition_info
+        The rewritten graph, and a mapping from unique nodes
+        in the new graph to associated partitioning information.
+    """
+    parents: defaultdict[IR, int] = defaultdict(int)
+    if _is_fusible_ir_node(ir, partition_info):
+        parents[ir] = 1
+    for node in traversal([ir]):
+        for child in node.children:
+            if _is_fusible_ir_node(child, partition_info):
+                parents[child] += 1
+    fusible = {node for node, count in parents.items() if count == 1}
+    state = {"fusible": fusible, "partition_info": partition_info}
+    mapper = CachingVisitor(_fuse_ir_node, state=state)
```

Comment:
OK, so this means that if we have:

```
HStack(
   HStack(
      SelectA,
      ...
   ),
   HStack(
      SelectA, 
      ...
   )
)
```
Then we will treat `SelectA` as _not_ fusible. Although it is a candidate for fusion, it has two different parents?

Reply:
Yeah, `HStack` can only have one child, but ignoring that for now: We now allow `SelectA` to be wrapped in a `Fused` (or `FusedIO`), but we mark the new object as `fusable=False` to block the parents of `SelectA` from also fusing to it.

To put it another way: We can (and do) wrap common-subplan nodes in `Fused`, but `Fused.fusable` must be set to `False` when these nodes are created.